### PR TITLE
Raise a square matrix to a power with #pow

### PR DESCRIPTION
**Pull request by v0dro**
Raise a square matrix to an integer power with #pow

tests for pow

raise to power

docs update

### CODE REVIEW

File: lib/nmatrix/math.rb [hunk 1]
```diff
@@ -364,6 +364,37 @@ def cov(opts={})
     deviation_scores.transpose.dot(deviation_scores) / denominator
   end
 
+  # Raise a square matrix to a power. Be careful of buffer overflows!
```
Comment (cjfuller) [inline lib/nmatrix/math.rb]: Do you mean "numeric overflow"?

File: lib/nmatrix/math.rb [hunk 2]
```diff
@@ -364,6 +364,37 @@ def cov(opts={})
...
+  # In case *n* is 0, a matrix of ones of the same dimension is returned. In case
+  # of negative *n*, the matrix is inverted and the absolute value of *n* taken 
+  # for computing the power.
+  # 
+  # == Arguments
+  # 
+  # * +n+ - Integer to which self is to be raised.
+  def pow n
+    raise ShapeError, "Only works with 2D square matrices." if shape[0] != shape[1]
```
Comment (cjfuller) [inline lib/nmatrix/math.rb]: You should also check that `shape.size == 2`.

File: lib/nmatrix/math.rb [hunk 3]
```diff
@@ -364,6 +364,37 @@ def cov(opts={})
...
+    raise TypeError, "Only works with integer powers" unless n.is_a?(Integer)
+
+    product = NMatrix.diag [1]*shape[0]
```
Comment (cjfuller) [inline lib/nmatrix/math.rb]: This means that this function always returns the default dtype and stype.  I think we probably want to preserve stype, and if n >= 0, preserve dtype, otherwise take the dtype as the more flexible of self and self.inverse's dtype.

File: lib/nmatrix/math.rb [hunk 4]
```diff
@@ -364,6 +364,37 @@ def cov(opts={})
...
+      return NMatrix.ones(shape)
+    elsif n < 0
+      n = n.abs
+      sequence.invert!
+    end
+
+    while n > 0
+      product = product.dot(sequence) if n % 2 == 1
+      n = n / 2
+      sequence = sequence.dot(sequence)
```
Comment (cjfuller) [inline lib/nmatrix/math.rb]: The `sequence = sequence.dot(sequence)` part is superfluous when n = 1

In general, this while loop doesn't feel very ruby-ish.  (It took me rather a while to figure out that what you're doing here is decomposing n into powers of two and then doing straight multiplication for what's left over.)  I probably would have had a separate function to do the decomposition of n, and then compute the appropriate powers.  This is probably marginally more efficient as you have it, though.  At minimum, can you put a comment here pointing out what this loop is doing?

File: spec/math_spec.rb [hunk 5]
```diff
@@ -377,6 +376,33 @@
...
+                                                 dtype: dtype))
+      end
+
+      it "raises a square matrix to odd power" do
+        expect(@n.pow(9)).to eq(NMatrix.new([4,4],[-275128,  279917, 176127, 237451,
+                                                  -260104,  394759, 166893,  296081,
+                                                  -704824,  285700, 186411,  262002,
+                                                  3209256,-1070870,-918741,-1318584],
+                                                  dtype: dtype))
+      end
```
Comment (cjfuller) [inline spec/math_spec.rb]: You're supporting negative powers and all stypes, so you need tests for negative powers and all combinations of stype/dtype.

**Comment by v0dro** [comment]
Ok will update in a while.

File: lib/nmatrix/math.rb (hunk 4 above)
Comment (v0dro) [inline lib/nmatrix/math.rb]: Yeah I also thought that its unrubyish but since its just 4 lines of code I really didn't bother about it.

How about I add a reference to this algorithm in the comments?

File: lib/nmatrix/math.rb (hunk 3 above)
Comment (v0dro) [inline lib/nmatrix/math.rb]: Ah yes my bad.

File: lib/nmatrix/math.rb (hunk 1 above)
Comment (v0dro) [inline lib/nmatrix/math.rb]: I mean a case where you assign 300 to a `char` in C and that doesnt do what you want it to.

That's numeric overflow?

**Comment by v0dro** [comment]
@cjfuller I've updated as per your suggestions.

Could you also check out #301? Its been due for a while...

File: lib/nmatrix/math.rb (hunk 3 above)
Comment (cjfuller) [inline lib/nmatrix/math.rb]: There's a dtype upcast:

```ruby
NMatrix.upcast(:int32, :int64)
=> :int64
```

File: lib/nmatrix/math.rb (hunk 1 above)
Comment (cjfuller) [inline lib/nmatrix/math.rb]: Yeah, I would have called that numeric overflow.  In my head buffer overflow is when you've allocated a buffer of, e.g., 10 chars, and you fill it up and then write to the 11th and maybe you segfault.  Not 100% sure if that's the standard definition of those terms, though?

Comment (mohawkjohn) [inline lib/nmatrix/math.rb]: Colin is correct.

File: lib/nmatrix/math.rb (hunk 3 above)
Comment (v0dro) [inline lib/nmatrix/math.rb]: What do I do if I want to upcast a whole bunch of dtypes to `int64`?

Something like `NMatrix.upcast([:byte, :int8,:int16,:int32], :int64)`. Is this possible?

Comment (cjfuller) [inline lib/nmatrix/math.rb]: `[:byte, :int8, :int16, :int32].map { |dtype| NMatrix.upcast(dtype, :int64) }`?

**Comment by v0dro** [comment]
@cjfuller could you please evaluate this PR?

**Comment by v0dro** [comment]
@cjfuller modified.

**Comment by v0dro** [comment]
Check #340
